controlpanel/api/cluster.py

```diff
@@ -124,17 +124,45 @@ def purge_user(user):
     helm.delete(f"init-user-{user.slug}")
 
 
-def create_app_role(name):
-    role = IAMRole(name)
-    role.save()
+class App:
+    """
+    Responsible for the apps-related interactions with the k8s cluster and AWS
+    """
 
+    APPS_NS = "apps-prod"
 
-def delete_app_role(name):
-    try:
-        aws.delete_role(name)
-    except Exception as e:
-        if not is_ignored_exception(e):
-            raise e
+    def __init__(self, app):
+        self.app = app
+
+    @property
+    def iam_role_name(self):
+        return f"{settings.ENV}_app_{self.app.slug}"
+
+    def create_iam_role(self):
+        role = IAMRole(self.iam_role_name)
+        role.save()
+
+    def delete_iam_role(self):
+        try:
+            aws.delete_role(self.iam_role_name)
+        except Exception as e:
+            if not is_ignored_exception(e):
+                raise e
+
+    @property
+    def url(self):
+        k8s = KubernetesClient(use_cpanel_creds=True)
+
+        repo_name = github_repository_name(self.app.repo_url)
+        ingresses = k8s.ExtensionsV1beta1Api.list_namespaced_ingress(
+            self.APPS_NS,
+            label_selector=f"repo={repo_name}",
+        ).items
+
+        if len(ingresses) != 1:
+            return ""
```

Comment:
Returning an empty string is fine as it evaluates to `False` in `{% if app_url %}`, but I think it makes more sense to return `None`

Reply:
Also, if there are more than one ingress, shouldn't we return one of them? It will be a valid URL, won't it?

Reply:
I'll have a think about the `None` but it may be better as `None` conveys better the "no URL" meaning.

Reply:
Regarding the multiple ingress I'd stick to return `""`/`None` as there should be only a single `Ingress` per app so if two are found we can't be 100% sure which URL is the right one.

I'd rather have the user come to us and ask why the URL is not shown and we then investigate than have shown a potentially wrong URL, the user freaks out, it comes to us anyway and we freak out as well.

It shouldn't happen anyway but I prefer to not pick a "random" URL when there is potential ambiguity.

Reply:
In my opinion, a working but "wrong" URL would be better than a missing one, but OK